pkg/controller/statefulset/stateful_set_control_test.go

```diff
@@ -2423,51 +2423,48 @@ type requestTracker struct {
 	err      error
 	after    int
 
-	parallelLock sync.Mutex
-	parallel     int
-	maxParallel  int
-
-	delay time.Duration
-}
-
-func (rt *requestTracker) errorReady() bool {
-	rt.Lock()
-	defer rt.Unlock()
-	return rt.err != nil && rt.requests >= rt.after
-}
-
-func (rt *requestTracker) inc() {
-	rt.parallelLock.Lock()
-	rt.parallel++
-	if rt.maxParallel < rt.parallel {
-		rt.maxParallel = rt.parallel
+	// this block should be updated consistently
+	parallelLock         sync.Mutex
+	parallelRequests     int
+	maxParallelRequests  int
+	parallelRequestDelay time.Duration
+}
+
+func (rt *requestTracker) trackParallelRequests() {
+	if rt.parallelLock.TryLock() {
+		// lock acquired: we are the only or the first concurrent request
+		// initialize the next set of parallel requests
+		rt.parallelRequests = 1
+	} else {
+		// lock is held by other requests
+		// now wait for the lock to increase the parallelRequests
+		rt.parallelLock.Lock()
+		rt.parallelRequests++
 	}
-	rt.parallelLock.Unlock()
-
-	rt.Lock()
-	defer rt.Unlock()
-	rt.requests++
-	if rt.delay != 0 {
-		time.Sleep(rt.delay)
+	defer rt.parallelLock.Unlock()
+	// update the local maximum of parallel collisions
+	if rt.maxParallelRequests < rt.parallelRequests {
+		rt.maxParallelRequests = rt.parallelRequests
+	}
+	// increase the chance of collisions
+	if rt.parallelRequestDelay != 0 {
+		time.Sleep(rt.parallelRequestDelay)
 	}
 }
 
-func (rt *requestTracker) reset() {
-	rt.parallelLock.Lock()
-	rt.parallel = 0
-	rt.parallelLock.Unlock()
-
-	rt.Lock()
-	defer rt.Unlock()
-	rt.err = nil
-	rt.after = 0
-	rt.delay = 0
-}
-
-func (rt *requestTracker) getErr() error {
+func (rt *requestTracker) incWithOptionalError() error {
 	rt.Lock()
 	defer rt.Unlock()
-	return rt.err
+	rt.requests++
+	if rt.err != nil && rt.requests >= rt.after {
+		// reset and pass the error
+		defer func() {
+			rt.err = nil
+			rt.after = 0
```

Comment:
`reset()` originally set rt.delay = 0. Why not do the same here for `rt.parallelRequestDelay`?

Reply:
As mentioned in https://github.com/kubernetes/kubernetes/pull/124624#issuecomment-2098186385 
this reset was never used before, because errors and parallel requests have not been tested at the same time before.

Logic of these two blocks `trackParallelRequests` and `incWithOptionalError` is independent of each other, so I think it makes sense to keep them separate.